src/toxcore/messenger/packet/msi.rs

```diff
@@ -0,0 +1,317 @@
+/*! Msi struct. Used by tox-av
+*/
+
+use nom::{ErrorKind, le_u8};
+use bitflags::*;
+
+use crate::toxcore::binary_io::*;
+
+/// Maximum size in bytes of msi message packet
+const MAX_MSI_PAYLOAD_SIZE: usize = 256;
+
+bitflags! {
+    /// Capabilities kind of msi packet. Used by bitwise OR.
+    pub struct CapabilitiesKind: u8 {
+        /// Send audio
+        const SEND_AUDIO = 4;
+        /// Send video
+        const SEND_VIDEO = 8;
+        /// Receive audio
+        const RECEIVE_AUDIO = 16;
+        /// Receive video
+        const RECEIVE_VIDEO = 32;
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct Capabilities(CapabilitiesKind);
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+/// Errors of msi session
+pub enum MsiErrorKind {
+    /// Error of none
+    MsiNone = 0,
+    /// Error of invalid message
+    InvalidMessage,
+    /// Error of invalid parameter
+    InvalidParam,
+    /// Error of invalid state
+    InvalidState,
+    /// Error of stray message
+    StrayMessage,
+    /// Error of system
+    System,
+    /// Error of handle
+    Handle,
+    /// Error of undisclosed
+    Undisclosed,
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct MsiError(MsiErrorKind);
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+/// Kind of msi request
+pub enum RequestKind {
+    /// Msi request of init
+    Init = 1,
+    /// Msi request of push
+    Push,
+    /// Msi request of pop
+    Pop,
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct Request(RequestKind);
+
+impl FromBytes for Request {
+    named!(from_bytes<Request>, do_parse!(
+        tag!("\x01") >>
+        verify!(le_u8, |size| size == 1) >>
+        value: call!(RequestKind::from_bytes) >>
+        (Request(value))
+    ));
+}
+
+impl ToBytes for Request {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        do_gen!(buf,
+            gen_le_u8!(0x01) >> // Request
+            gen_le_u8!(0x01) >> // Size
+            gen_le_u8!(self.0 as u8)
+        )
+    }
+}
+
+impl FromBytes for MsiError {
+    named!(from_bytes<MsiError>, do_parse!(
+        tag!("\x02") >>
+        verify!(le_u8, |size| size == 1) >>
+        value: call!(MsiErrorKind::from_bytes) >>
+        (MsiError(value))
+    ));
+}
+
+impl ToBytes for MsiError {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        do_gen!(buf,
+            gen_le_u8!(0x02) >> // MsiError
+            gen_le_u8!(0x01) >> // Size
+            gen_le_u8!(self.0 as u8)
+        )
+    }
+}
+
+impl FromBytes for CapabilitiesKind {
+    named!(from_bytes<CapabilitiesKind>, do_parse!(
+        value: le_u8 >>
+        (CapabilitiesKind {
+            bits: value,
+        })
+    ));
+}
+
+impl FromBytes for Capabilities {
+    named!(from_bytes<Capabilities>, do_parse!(
+        tag!("\x03") >>
+        verify!(le_u8, |size| size == 1) >>
```

Comment:
It seems it's a const? Then why not put it together with tag? `tag!("\x03\x01")`

Reply:
because for now it is const

Reply:
But then we will have to change this anyway? Remove `verify` and use the size.

Reply:
Because it describes the proto better than putting 0x01 along with tag